### PR TITLE
Display preview feature for Bam datatype

### PR DESCRIPTION
**Pull request by ashvark**
This PR has the implementation of **display_data** function for **BAM** dataype. It enables the users to view the content of the Bam file with in the galaxy itself.

### CODE REVIEW

File: lib/galaxy/datatypes/binary.py [hunk 1]
```diff
@@ -462,6 +469,53 @@ def to_archive(self, trans, dataset, name=""):
...
+    def get_chunk(self, trans, dataset, offset=0, ck_size=None):
+        bamfile = pysam.AlignmentFile(dataset.file_name, "rb")
+        ck_size = 100 # 100 lines
+        ck_data=""
+        lineNumber=0
+        if(offset == 0):
+            ck_data = bamfile.text
+        for f in bamfile.fetch(until_eof=True):
+            lineNumber+=1
+            if (lineNumber > offset and lineNumber <= (offset + ck_size)):
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py]: The time/cpu spent to reach the lines increases with every time the users displays more lines.
You can replace the line number approach for specifying the offset with `bamfile.seek()` and `bamfile.tell()` methods (which weirdly enough is not documented in http://pysam.readthedocs.io/en/latest/api.html). So after fetching the number of alignments that you want (maybe 1000 is more reasonable than 100?), set `offset` to `bamfile.tell()` and before iterating over alignments jump to the correct offset (if it is not 0) using `bamfile.seek(offset)`

File: lib/galaxy/datatypes/binary.py [hunk 2]
```diff
@@ -236,6 +239,10 @@ class Bam( Binary ):
     MetadataElement( name="reference_lengths", default=[], desc="Chromosome Lengths", param=MetadataParameter, readonly=True, visible=False, optional=True, no_value=[] )
     MetadataElement( name="bam_header", default={}, desc="Dictionary of BAM Headers", param=MetadataParameter, readonly=True, visible=False, optional=True, no_value={} )
 
+    def __init__(self, **kwd):
+        """Initialize taxonomy datatype"""
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py]: That is not the correct docstring.

File: lib/galaxy/datatypes/binary.py [hunk 3]
```diff
@@ -462,6 +469,53 @@ def to_archive(self, trans, dataset, name=""):
         file_paths.append(dataset.metadata.bam_index.file_name)
         return zip(file_paths, rel_paths)
 
+
+    def get_chunk(self, trans, dataset, offset=0, ck_size=None):
+        bamfile = pysam.AlignmentFile(dataset.file_name, "rb")
+        ck_size = 100 # 100 lines
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py]: You need to have 2 spaces before inlines comments

File: lib/galaxy/datatypes/binary.py [hunk 4]
```diff
@@ -462,6 +469,53 @@ def to_archive(self, trans, dataset, name=""):
...
+        ck_data=""
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py]: You need single spaces around operators.

File: lib/galaxy/datatypes/binary.py [hunk 5]
```diff
@@ -462,6 +469,53 @@ def to_archive(self, trans, dataset, name=""):
...
+        lineNumber=0
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py]: We are not using CamelCase for variables, this should probably be `line_number = 0`

File: lib/galaxy/datatypes/binary.py [hunk 6]
```diff
@@ -462,6 +469,53 @@ def to_archive(self, trans, dataset, name=""):
...
+        if(offset == 0):
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py]: the brackets are unnecessary, this should be `if offset == 0:`.

File: lib/galaxy/datatypes/binary.py [hunk 7]
```diff
@@ -462,6 +469,53 @@ def to_archive(self, trans, dataset, name=""):
...
+            ck_data = bamfile.text
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py]: We probably want to get the header only in the first chunk (i.e if offset is 0), since the header can be multiple thousand lines long.

File: lib/galaxy/datatypes/binary.py [hunk 8]
```diff
@@ -462,6 +469,53 @@ def to_archive(self, trans, dataset, name=""):
...
+        for f in bamfile.fetch(until_eof=True):
+            lineNumber+=1
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py]: Since it's probably better to use the offset you can do `for line_number, f in enumerate(bamfile):`.
Also f is frequently use to indicate a file handle, in this situation it would probably be better to rename `f` to `read`, `alignment` or `segment`.

File: lib/galaxy/datatypes/binary.py [hunk 9]
```diff
@@ -462,6 +469,53 @@ def to_archive(self, trans, dataset, name=""):
         file_paths.append(dataset.metadata.bam_index.file_name)
         return zip(file_paths, rel_paths)
 
+
+    def get_chunk(self, trans, dataset, offset=0, ck_size=None):
+        bamfile = pysam.AlignmentFile(dataset.file_name, "rb")
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py]: you probably want to use `with pysam.AlignmentFile(dataset.file_name, "rb") as bamfile:` here.

File: lib/galaxy/datatypes/binary.py [hunk 10]
```diff
@@ -462,6 +469,53 @@ def to_archive(self, trans, dataset, name=""):
...
+            lineNumber+=1
+            if (lineNumber > offset and lineNumber <= (offset + ck_size)):
+                bamline = f.tostring(bamfile)
+                # Galaxy display each tag as separate column because 'tostring()' funcition put spaces in between each tag of tags column. 
+                # Below code will remove spaces between each tag. 
+                bamlineModified = ('\t').join(bamline.split()[:11] + [('').join(bamline.split()[11:])])
+                ck_data=ck_data +"\n" + bamlineModified
+            elif (lineNumber > (offset + ck_size)):
+                break
+        last_read = offset + ck_size
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py]: here you can set a new offset using `bamfile.tell()`

File: lib/galaxy/datatypes/binary.py [hunk 11]
```diff
@@ -462,6 +469,53 @@ def to_archive(self, trans, dataset, name=""):
...
+                        'offset': last_read } )
+
+    def display_data( self, trans, dataset, preview=False, filename=None, to_ext=None, offset=None, ck_size=None, **kwd):
+        preview = util.string_as_bool( preview )
+        if offset is not None:
+            return self.get_chunk(trans, dataset, offset, ck_size)
+        elif to_ext or not preview:
+            return super( Bam, self ).display_data( trans, dataset, preview, filename, to_ext, **kwd )
+        else:
+            column_names = 'null'
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py]: I think `''` would be better as a fallback.

File: lib/galaxy/datatypes/binary.py [hunk 12]
```diff
@@ -462,6 +469,53 @@ def to_archive(self, trans, dataset, name=""):
...
+            if dataset.metadata.column_names:
+                column_names = dataset.metadata.column_names
+            elif hasattr(dataset.datatype, 'column_names'):
+                column_names = dataset.datatype.column_names
+            column_types = dataset.metadata.column_types
+            if not column_types:
+                column_types = []
+            column_number = dataset.metadata.columns
+            if column_number is None:
+                column_number = 'null'
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py]: it may be more accurate to set this to 1?

**Review by mvdbeek — CHANGES_REQUESTED** [review]
This is a promising start and could be quite useful for some applications!

**Comment by ashvark** [comment]
@mvdbeek Thank you very much for the review. I managed to do all the changes except the bamfile.seek() functionality. I could not iIterate the bamfile after i use seek() whereas I could do without seek() . 

Below is  the snapshot. Can you help me ? 

image

**Comment by mvdbeek** [comment]
You really need to use the offset that you get from `.tell()`, just specifying `3` is not likely to work.
See this as an example:
```
In [28]: with pysam.AlignmentFile('/Users/mvandenb/src/readtagger/test/data/cornercase.bam') as f:
    ...:     first_offset = f.tell()
    ...:     print(first_offset)
    ...:     r = f.next()
    ...:     print(r.query_name)
    ...:     print(f.tell())
    ...:     next_read = f.next()
    ...:     print(next_read.query_name)
    ...:     print(f.tell())
    ...:     f.seek(first_offset)
    ...:     first_read_seek = f.next()
    ...:     print(first_read_seek.query_name == r.query_name)
    ...:
1442119680
HWI-D00405:129:C6KNAANXX:3:2210:9430:87369
1442120130
HWI-D00405:129:C6KNAANXX:4:2103:18945:51370
1474232320
True
```

File: lib/galaxy/datatypes/binary.py [hunk 13]
```diff
@@ -462,6 +468,52 @@ def to_archive(self, trans, dataset, name=""):
...
+    def get_chunk(self, trans, dataset, offset=0, ck_size=None):
+        index_file = dataset.metadata.bam_index
+        with pysam.AlignmentFile(dataset.file_name, "rb", index_filename=index_file.file_name) as bamfile:
+            ck_size = 1000  # 1000 lines
+            ck_data = ""
+            line_number = 0
+            if offset == 0:
+                ck_data = bamfile.text
+            for line_number, alignment,  in enumerate(bamfile):
+                if line_number > offset and line_number <= (offset + ck_size):
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py]: you probably want to change this to `if line_number < ck_size:`

File: lib/galaxy/datatypes/binary.py [hunk 14]
```diff
@@ -462,6 +468,52 @@ def to_archive(self, trans, dataset, name=""):
...
 
+    def get_chunk(self, trans, dataset, offset=0, ck_size=None):
+        index_file = dataset.metadata.bam_index
+        with pysam.AlignmentFile(dataset.file_name, "rb", index_filename=index_file.file_name) as bamfile:
+            ck_size = 1000  # 1000 lines
+            ck_data = ""
+            line_number = 0
+            if offset == 0:
+                ck_data = bamfile.text
+            for line_number, alignment,  in enumerate(bamfile):
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py]: you can drop the `, ` after alignment

File: lib/galaxy/datatypes/binary.py (hunk 13 above)
Comment (mvdbeek) [reply lib/galaxy/datatypes/binary.py]: Actually, we can switch this around and break here if the `line_number` exceeds `ck_size`:
```
                if line_number > ck_size:
                    break
```

File: lib/galaxy/datatypes/binary.py [hunk 15]
```diff
@@ -462,6 +468,52 @@ def to_archive(self, trans, dataset, name=""):
...
+                if line_number > offset and line_number <= (offset + ck_size):
+                    bamline = alignment.tostring(bamfile)
+                    # Galaxy display each tag as separate column because 'tostring()' funcition put spaces in between each tag of tags column. 
+                    # Below code will remove spaces between each tag. 
+                    bamline_modified = ('\t').join(bamline.split()[:11] + [('').join(bamline.split()[11:])])
+                    ck_data = ck_data +"\n" + bamline_modified
+                elif line_number > (offset + ck_size):
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py]: you can drop this if you implement https://github.com/galaxyproject/galaxy/pull/4279/files#r125709901

File: lib/galaxy/datatypes/binary.py [hunk 16]
```diff
@@ -462,6 +468,52 @@ def to_archive(self, trans, dataset, name=""):
...
+                    break
+        last_read = offset + ck_size
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py]: I'd change this to `offset = bamfile.tell()`

File: lib/galaxy/datatypes/binary.py [hunk 17]
```diff
@@ -462,6 +468,52 @@ def to_archive(self, trans, dataset, name=""):
...
+        return dumps( { 'ck_data': util.unicodify( ck_data ),
+                        'offset': last_read } )
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py]: and then change `last_read` to `offset`

File: lib/galaxy/datatypes/sniff.py [hunk 18]
```diff
@@ -478,7 +477,7 @@ def handle_uploaded_dataset_file( filename, datatypes_registry, ext='auto', is_m
         ext = guess_ext( filename, sniff_order=datatypes_registry.sniff_order, is_multi_byte=is_multi_byte )
 
     if check_binary( filename ):
-        if not Binary.is_ext_unsniffable(ext) and not datatypes_registry.get_datatype_by_extension( ext ).sniff( filename ):
+        if not galaxy.datatypes.binary.is_ext_unsniffable(ext) and not datatypes_registry.get_datatype_by_extension( ext ).sniff( filename ):
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/sniff.py]: Is this change necessary?

Comment (ashvark) [reply lib/galaxy/datatypes/sniff.py]: Compilation breaks when I have both **'from galaxy.datatypes.binary import Binary'** in **sniff.py** and **'from galaxy.datatypes.tabular import Sam'** in **binary.py** . I donot understand the reason behind that. Code is compiling properly only when i remove **'from galaxy.datatypes.binary import Binary'**  from **sniff.py**.

File: lib/galaxy/datatypes/sniff.py [hunk 19]
```diff
@@ -487,7 +487,7 @@ def handle_uploaded_dataset_file( filename, datatypes_registry, ext='auto', is_m
         ext = guess_ext( filename, sniff_order=datatypes_registry.sniff_order, is_multi_byte=is_multi_byte )
 
     if check_binary( filename ):
-        if not Binary.is_ext_unsniffable(ext) and not datatypes_registry.get_datatype_by_extension( ext ).sniff( filename ):
+        if not datatypes.binary.Binary.is_ext_unsniffable(ext) and not datatypes_registry.get_datatype_by_extension( ext ).sniff( filename ):
```
Comment (ashvark) [inline lib/galaxy/datatypes/sniff.py]: I would like to know whether `datatypes.binary.Binary.is_ext_unsniffable(ext)` is the right way to call this function ?

Comment (mvdbeek) [reply lib/galaxy/datatypes/sniff.py]: No, I don't think you should change sniff.py.
I think you need very few things from the Sam class.
In fact I think you only need the column names, so you could just initialize them in the Bam class.

Comment (ashvark) [reply lib/galaxy/datatypes/sniff.py]: @mvdbeek  I already tried initializing the column_names with out importing Sam class. But column names are not displayed in the galaxy app. Then when I inherited Sam for Bam, colmun names started populating. I donot understand the reason behind.. i think initializing column names alone in Bam class is not enough.

**Comment by mvdbeek** [comment]
@galaxybot test this

File: lib/galaxy/datatypes/binary.py [hunk 20]
```diff
@@ -235,6 +237,13 @@ class Bam( Binary ):
     MetadataElement( name="reference_names", default=[], desc="Chromosome Names", param=MetadataParameter, readonly=True, visible=False, optional=True, no_value=[] )
     MetadataElement( name="reference_lengths", default=[], desc="Chromosome Lengths", param=MetadataParameter, readonly=True, visible=False, optional=True, no_value=[] )
     MetadataElement( name="bam_header", default={}, desc="Dictionary of BAM Headers", param=MetadataParameter, readonly=True, visible=False, optional=True, no_value={} )
+    MetadataElement( name="columns", default=0, desc="Number of columns", readonly=True, visible=False, no_value=0 )
+    MetadataElement( name="column_types", default=[], desc="Column types", param=metadata.ColumnTypesParameter, readonly=True, visible=False, no_value=[] )
+    MetadataElement( name="column_names", default=[], desc="Column names", readonly=True, visible=False, optional=True, no_value=[] )
+
```
Comment (ashvark) [inline lib/galaxy/datatypes/binary.py]: It seems that above three lines are mandatory inorder to show the column names in galaxy

File: lib/galaxy/datatypes/binary.py [hunk 21]
```diff
@@ -423,6 +432,8 @@ def set_meta( self, dataset, overwrite=True, **kwd ):
             dataset.metadata.read_groups = [ read_group['ID'] for read_group in dataset.metadata.bam_header.get( 'RG', [] ) if 'ID' in read_group ]
             dataset.metadata.sort_order = dataset.metadata.bam_header.get( 'HD', {} ).get( 'SO', None )
             dataset.metadata.bam_version = dataset.metadata.bam_header.get( 'HD', {} ).get( 'VN', None )
+            dataset.metadata.columns = 12
+            dataset.metadata.column_types = ['str', 'int', 'str', 'int', 'int', 'str', 'str', 'int', 'int', 'str', 'str', 'str']
```
Comment (ashvark) [inline lib/galaxy/datatypes/binary.py]: Could not understand the importance of above two lines. Column names are displayed even if i removing these two lines.

File: lib/galaxy/datatypes/binary.py [hunk 22]
```diff
@@ -462,6 +473,58 @@ def to_archive(self, trans, dataset, name=""):
...
 
+    def get_chunk( self, trans, dataset, offset=0, ck_size=None ):
+        index_file = dataset.metadata.bam_index
+        with pysam.AlignmentFile( dataset.file_name, "rb", index_filename=index_file.file_name ) as bamfile:
+            ck_size = 300  # 300 lines
+            ck_data = ""
+            last_read = 0
+            header_line_count = 0
+            if offset == 0:
+                ck_data = bamfile.text
```
Comment (ashvark) [inline lib/galaxy/datatypes/binary.py]: @mvdbeek Like you mentioned already, What happens when bam file has several lines of header. Do I need to trim if it exceeds certain limit ?

File: lib/galaxy/datatypes/binary.py [hunk 23]
```diff
@@ -462,6 +473,58 @@ def to_archive(self, trans, dataset, name=""):
...
+
+    def display_data( self, trans, dataset, preview=False, filename=None, to_ext=None, offset=None, ck_size=None, **kwd):
+        preview = util.string_as_bool( preview )
+        if offset is not None:
+            return self.get_chunk( trans, dataset, offset, ck_size )
+        elif to_ext or not preview:
+            return super( Bam, self ).display_data( trans, dataset, preview, filename, to_ext, **kwd )
+        else:
+            column_names = ''
+            if dataset.metadata.column_names:
```
Comment (ashvark) [inline lib/galaxy/datatypes/binary.py]: **dataset.metadata.column_names** will be always empty because I am not setting anything in the set_meta() function. Even in the Sam class I could see the same thing.

Comment (mvdbeek) [reply lib/galaxy/datatypes/binary.py]: yes, I think you could set the column_names `MetadataElement` directly to the correct column_names, since they are not going to change for this datatype. Same for the column_number and column_types, they're always going to be the same, so you can just fix them in the `MetadataElement`.

File: lib/galaxy/datatypes/binary.py (hunk 22 above)
Comment (mvdbeek) [reply lib/galaxy/datatypes/binary.py]: I think for now we should cut them off as you're doing now. In the future we could work with the `comment_lines` MetadataElement and be a little smarter about this, i.e let the user choose to see the full header. I don't think this is too important for the start.

File: lib/galaxy/datatypes/binary.py [hunk 24]
```diff
@@ -462,6 +473,58 @@ def to_archive(self, trans, dataset, name=""):
...
+                # FIXME: Can be problematic if bam has million lines of header
+                if ( line_number + header_line_count ) > ck_size:
+                    break
+                else:
+                    bamline = alignment.tostring( bamfile )
+                    # Galaxy display each tag as separate column because 'tostring()' funcition put spaces in between each tag of tags column.
+                    # Below code will remove spaces between each tag.
+                    bamline_modified = ('\t').join( bamline.split()[:11] + [ ('').join(bamline.split()[11:]) ] )
+                    ck_data = ck_data + "\n" + bamline_modified
+                last_read = bamfile.tell()
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py]: you can pull this out of the for loop, you only need to set this once, plus this prevents an undefined variable if you break out due to too many header lines. Also, could you change the variable name to `offset` ?

File: lib/galaxy/datatypes/binary.py (hunk 21 above)
Comment (mvdbeek) [reply lib/galaxy/datatypes/binary.py]: I think (I could be wrong) the main purpose is for accessing this sort of metadata in the tool wrappers without needing to call `str()` or `int()`. But since those are always the same I would prefer setting this directly as the default in the MetadataElement.

**Comment by mvdbeek** [comment]
@galaxybot test this

File: lib/galaxy/datatypes/binary.py (hunk 24 above)
Comment (ashvark) [reply lib/galaxy/datatypes/binary.py]: if i put the ` last_read = bamfile.tell()` , I am missing one line form bamfile for every chunk. Because, I am breaking the loop only after enumerating.

Comment (mvdbeek) [reply lib/galaxy/datatypes/binary.py]: Ah, right. Then you should make this the first line of the loop, otherwise `last_read` will be undefined (or if you rename `last_read` to `offset` `offset` would always be 0) if the header is larger than `ck_size`.

File: lib/galaxy/datatypes/binary.py [hunk 25]
```diff
@@ -462,6 +469,56 @@ def to_archive(self, trans, dataset, name=""):
...
+            header_line_count = 0
+            if offset == 0:
+                ck_data = bamfile.text
+                header_line_count = bamfile.text.count('\n')
+            else:
+                bamfile.seek( offset )
+            for line_number, alignment in enumerate( bamfile ) :
+                # return only Header lines if 'header_line_count' exceeds 'ck_size'
+                # FIXME: Can be problematic if bam has million lines of header
+                last_read = bamfile.tell()
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py]: Can you rename this to offset ? `last_read` is a little confusing here.

File: lib/galaxy/datatypes/binary.py [hunk 26]
```diff
@@ -462,6 +469,56 @@ def to_archive(self, trans, dataset, name=""):
...
+                if ( line_number + header_line_count ) > ck_size:
+                    break
+                else:
+                    bamline = alignment.tostring( bamfile )
+                    # Galaxy display each tag as separate column because 'tostring()' funcition put spaces in between each tag of tags column.
+                    # Below code will remove spaces between each tag.
+                    bamline_modified = ('\t').join( bamline.split()[:11] + [ ('').join(bamline.split()[11:]) ] )
+                    ck_data = ck_data + "\n" + bamline_modified
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py]: I'd prefer `"%s\n%s" % (ck_data, bamline_modified)`, this is how we usually handle string concatenation in galaxy.

File: lib/galaxy/datatypes/binary.py [hunk 27]
```diff
@@ -462,6 +469,56 @@ def to_archive(self, trans, dataset, name=""):
...
+
+    def display_data( self, trans, dataset, preview=False, filename=None, to_ext=None, offset=None, ck_size=None, **kwd):
+        preview = util.string_as_bool( preview )
+        if offset is not None:
+            return self.get_chunk( trans, dataset, offset, ck_size )
+        elif to_ext or not preview:
+            return super( Bam, self ).display_data( trans, dataset, preview, filename, to_ext, **kwd )
+        else:
+            column_names = dataset.metadata.column_names
+            if not column_names:
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py:504]: I don't think you need to do this fallback, `dataset.metadata.column_names` should always be set (I haven't verified that).

File: lib/galaxy/datatypes/binary.py (hunk 26 above)
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py]: I'd prefer "%s\n%s" % (ck_data, bamline_modified), this is how we usually handle string concatenation in galaxy.

**Comment by mvdbeek** [comment]
@galaxybot test this

File: lib/galaxy/datatypes/binary.py (hunk 27 above)
Comment (mvdbeek) [reply lib/galaxy/datatypes/binary.py:504]: Yep, this is what MetadataElement(name='column_names') does. I think it's clearer if we remove the `if not ...` statements.

File: lib/galaxy/datatypes/binary.py [hunk 28]
```diff
@@ -462,6 +467,55 @@ def to_archive(self, trans, dataset, name=""):
...
         return zip(file_paths, rel_paths)
 
+    def get_chunk( self, trans, dataset, offset=0, ck_size=None ):
+        index_file = dataset.metadata.bam_index
+        with pysam.AlignmentFile( dataset.file_name, "rb", index_filename=index_file.file_name ) as bamfile:
+            ck_size = 300  # 300 lines
+            ck_data = ""
+            header_line_count = 0
+            if offset == 0:
+                ck_data = bamfile.text
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py]: what do you think about doing `ck_data = bamfile.text.replace('\t', ' ')` ? That would prevent the header lines to stretch the columns too far (It is a dirty hack ... It would be better to set and render the comment lines in a single cell)

Comment (ashvark) [reply lib/galaxy/datatypes/binary.py]: yes, i think it is good idea. I will update my code.

**Comment by ashvark** [comment]
image

@mvdbeek  I could not understand why 'continuous-integration/travis-ci/pr' failed. can you help me ?

**Comment by nsoranzo** [comment]
@ashvark I suppose that was a temporary issue, I've restarted the failed builds.

**Comment by ashvark** [comment]
@nsoranzo thank you very much

**Comment by jmchilton** [comment]
@galaxybot test this

**Comment by ashvark** [comment]
I could not understand why toolshed test has failed status eventhough all the tests are passed.

**Comment by bgruening** [comment]
@ashvark everything looks ok now, the tests pass.

**Comment by martenson** [comment]
@mvdbeek would you mind re-visiting your review?

**Comment by mvdbeek** [comment]
Yeah, it's OK.

**Comment by mvdbeek** [comment]
Thanks a lot @ashvark, this is great!

**Comment by ashvark** [comment]
Thank you all for your help in implementing this feature.

File: lib/galaxy/datatypes/binary.py [hunk 29]
```diff
@@ -462,6 +467,55 @@ def to_archive(self, trans, dataset, name=""):
         file_paths.append(dataset.metadata.bam_index.file_name)
         return zip(file_paths, rel_paths)
 
+    def get_chunk( self, trans, dataset, offset=0, ck_size=None ):
+        index_file = dataset.metadata.bam_index
```
Comment (jmchilton) [inline lib/galaxy/datatypes/binary.py:471]: @mvdbeek Do you happen to know if this function has the potential to consume unbounded memory if the bam file is problematic or is it going to always be pretty managable despite the bam file?

Comment (mvdbeek) [reply lib/galaxy/datatypes/binary.py:471]: Yes, bam lines (reads) can become huge. If someone did a genome-against-genome alignment I think this could be problematic.

Comment (jmchilton) [reply lib/galaxy/datatypes/binary.py:471]: We verified the problem on main wasn't data loaded into peak attributes - sorry for the false alarm. Maybe we should create an issue for known points in Galaxy where the amount of data loaded is unbounded, I don't know.
